sympy/polys/matrices/lll.py

```diff
@@ -0,0 +1,85 @@
+from math import floor as mfloor
+from typing import Tuple
+
+from sympy.polys.domains import ZZ, QQ
+from sympy.polys.matrices.exceptions import DMRankError, DMShapeError, DMValueError, DMDomainError
+
+logic_error_template = '''FAILURE IN ALGORITHM LOGIC!!!!
+Please file a bug report at https://github.com/sympy/sympy and include this input:
+{}'''
+linear_dependent_error = "input matrix contains linearly dependent rows"
+
+
+def ddm_lll(x, delta=QQ(3, 4)):
+    if QQ(1, 4) >= delta or delta > QQ(1, 1):
+        DMValueError("delta must lie in range (0.25, 1)")
+    if x.shape[0] > x.shape[1]:
+        DMShapeError("input matrix must have shape (m, n) with m <= n")
+    if x.domain != ZZ:
+        DMDomainError("input matrix domain must be ZZ")
+    m = x.shape[0]
+    n = x.shape[1]
+    k = 1
+    y = x.copy()
+    y_star = x.zeros((m, n), QQ)
+    mu = x.zeros((m, m), QQ)
+    g_star = [QQ(0, 1) for _ in range(m)]
+    half = QQ(1, 2)
+
+    def closest_integer(x):
+        return ZZ(mfloor(x + half))
+
+    def lovasz_condition(k: int) -> bool:
+        return g_star[k] >= ((delta - pow(mu[k][k - 1], 2)) * g_star[k - 1])
+
+    def mu_small(k: int, j: int) -> bool:
+        return abs(mu[k][j]) <= half
+
+    def dot_rows(x, y, rows: Tuple[int, int]):
+        return sum([x[rows[0]][z] * y[rows[1]][z] for z in range(x.shape[1])])
+
+    def reduce_row(mu, y, rows: Tuple[int, int]):
+        r = closest_integer(mu[rows[0]][rows[1]])
+        y[rows[0]] = [y[rows[0]][z] - r * y[rows[1]][z] for z in range(n)]
+        for j in range(rows[1]):
+            mu[rows[0]][j] -= r * mu[rows[1]][j]
+        mu[rows[0]][rows[1]] -= r
+
+    for i in range(m):
+        y_star[i] = [QQ.convert_from(z, ZZ) for z in y[i]]
+        for j in range(i):
+            try:
+                mu[i][j] = dot_rows(y, y_star, (i, j)) / g_star[j]
+            except ZeroDivisionError:
+                raise DMRankError(linear_dependent_error)
```

Comment:
Again this try/except encloses more code than it should because it calls a function and performs a division both inside the try block. To catch the division in the right place it would be:
```python
dot_prod = dot_rows(y, y_star, (i, j)) 
try:
    mu[i][j] = dot_prod / g_star[j]
except ZeroDivisionError:
    raise DMRankError(linear_dependent_error)
```
Even then the indexing like `g_star[j]` is in principle executing arbitrary code. Here it isn't because we use list but if this was a Matrix or something then it would.

Really though it's better and clearer just to use an `if`:
```python
if g_star[j] == 0:
    raise DMRankError(...)
mu[i][j] = dot_rows(y, y_star, (i, j)) / g_star[j]
```
Then there's no chance of ever catching the wrong exception and the condition under which we would raise here is made explicit.

Reply:
Ok, I am adding this in too.